Initial implementation of SideChannelDevice

arjan-s opened the pull request:
This pull request for https://github.com/django-otp/django-otp/issues/32 is not really meant to be merged, but to ask for reviews. Hopefully at some point it will be good enough to be merged.

psagers commented on src/django_otp/models.py:
Strictly speaking, the previous conditional will still succeed for very short time after this statement. Maybe use a `>` test to remove the race condition?

arjan-s replied:
Thank you, I missed that. It's fixed now.

psagers commented on src/django_otp/models.py:
Similar to ModelForm, ThrottlingMixin automatically commits in situations like this, unless given `commit=False`. Might be more consistent.

arjan-s replied:
Good idea, it's better to be consistent.

psagers commented on src/django_otp/plugins/otp_email/models.py:
```diff
@@ -19,30 +15,18 @@ def key_validator(value):
     def generate_challenge(self):
-        token = totp(self.bin_key)
-        body = render_to_string('otp/email/token.txt', {'token': token})
+        self.generate_token()
```
This whole thing came out of a request for customizing the valid time of email tokens. Feel free to fold in a `OTP_EMAIL_TOKEN_VALID_TIME` or similar setting here if you like.

arjan-s replied:
Thanks, I've added a new setting called `OTP_EMAIL_TOKEN_VALID_SECS`. I chose `SECS` instead of `TIME` to make the time unit unambiguous.

arjan-s commented:
I addressed the review comments and updated the documentation to include `SideChannelDevice`. All tests are still passing.
Note: I'll test this PR more extensively in a few days.

arjan-s commented:
Okay, I've tested the changes in this PR in a larger project and found everything to be working correctly. At the moment I don't see more to do, so I'll pass it back to @psagers for review. Please let me know if there is more to do.

psagers commented:
Do we at least have a test case for rejecting an email token after it's been used? Note that the test environment already includes [freezegun](https://pypi.org/project/freezegun/) for deterministic time management.

arjan-s commented:
Good point, I now added multiple tests for the new implementation of EmailDevice.

PritamDutt commented:
Thank you for completing this :)
